fix(engine): allow turning off implicit variable update detection

juja0 opened the pull request:
- adds a config property to allow disabling detection of implicit variable updates

related to #2969

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/camunda/camunda-bpm-platform?pullRequest=3008) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/camunda/camunda-bpm-platform?pullRequest=3008) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/camunda/camunda-bpm-platform?pullRequest=3008) it.</sub>

ThorbenLindhauer commented on engine/src/main/java/org/camunda/bpm/engine/impl/persistence/entity/util/TypedValueField.java:
```diff
@@ -152,8 +152,12 @@ protected boolean isMutableValue(TypedValue value) {
+
   protected boolean isValuedImplicitlyUpdated() {
-    if (cachedValue != null && isMutableValue(cachedValue)) {
+    if (cachedValue != null && isImplicitVariableUpdateDetectionEnabled() && isMutableValue(cachedValue)) {
```
I suggest to move this condition to `#notifyImplicitValueUpdate`. This seems a little more fitting to me, as `#isValuedImplicitlyUpdated` sounds like it returns the value independent of configuration. Maybe rename `notifyImplicitValueUpdate` to `notifyImplicitValueUpdateIfEnabled` if we want to be most precise.

juja0 replied:
Sounds good. Will do.

ThorbenLindhauer commented on engine/src/test/java/org/camunda/bpm/engine/impl/persistence/entity/util/TypedValueFieldTest.java:
```diff
@@ -0,0 +1,167 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TypedValueFieldTest {
```
I'd like to suggest writing tests that use only the API. That makes it more robust against future refactorings (and also us more certain that we are actually fixing the behavior from the user perspective).

You can test the case with disabled implicit updates as follows: Implement a serializable POJO class and implement `writeObject` (see 3.3 here: https://www.baeldung.com/java-serialization) that, depending on the value of a static flag, throws an exception. Then in the test, disable the flag when you create the variable initially and enable it before you retrieve it.

Lastly, you can move the test next to `org.camunda.bpm.engine.test.api.variables.ImplicitVariableUpdateTest`, e.g. call it `ImplicitVariableUpdateDisabledTest`.

juja0 replied:
This looks good. Will update the PR with the suggested changes sometime this week.

ThorbenLindhauer replied:
That's great, thank you.

ThorbenLindhauer commented on engine/src/test/java/org/camunda/bpm/engine/impl/persistence/entity/util/TypedValueFieldTest.java:
```diff
@@ -0,0 +1,167 @@
+
+  @Before
+  public void init() {
+    this.configuration = (ProcessEngineConfigurationImpl) ProcessEngineConfiguration.createStandaloneInMemProcessEngineConfiguration();
```
You can use `ProcessEngineBootstrapRule` to create a new engine. For example, have a look at https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/test/java/org/camunda/bpm/engine/test/api/history/BulkHistoryDeleteDmnDisabledTest.java#L47-L49 how to use it. Also, you can declare it as a class rule that way, so that we avoid the costly operation of building a new engine as much as possible.

juja0 replied:
Will do.

ThorbenLindhauer reviewed (changes requested):
Thank you for preparing a PR! The functionality is good and I think we can make the test more robust by relying more on API methods, please have a look at the detail comments.

Could you please also make a PR to the docs? In particular, you could add the new configuration to https://github.com/camunda/camunda-docs-manual/blob/master/content/reference/deployment-descriptors/tags/process-engine.md.

juja0 commented:
Removed usage of internal apis and updated the test based on what we discussed. 

Will create separate PR for doc update.

I also noticed that the existing test `ImplicitVariableUpdateTest.testUpdate` detects implicit updates in a more succinct way.

In light of that, I wasn't sure if the serialization based check that we discussed previously is still applicable so I've included both for the disabled case and added the serialization based check to the enabled case (in `ImplicitVariableUpdateTest`).

ThorbenLindhauer commented:
Thanks a lot for your contribution. It turned out nicely and I merged it via https://github.com/camunda/camunda-bpm-platform/commit/5d3e0cfd48d246111305c44b56901e732f2dce17. I will check the docs PR next.
